src/mappings/horizonStaking.ts

```diff
@@ -72,3 +83,135 @@ export function handleHorizonStakeWithdrawn(event: HorizonStakeWithdrawn): void
+    let indexer = Indexer.load(event.params.serviceProvider.toHexString())!
+    let provision = createOrLoadProvision(event.params.serviceProvider, event.params.verifier, event.block.timestamp)
+
+    indexer.thawingTokens = indexer.thawingTokens.plus(event.params.tokens)
+    indexer.save()
+
+    dataService.totalTokensThawing = dataService.totalTokensThawing.plus(event.params.tokens)
+    dataService.save()
+
+    graphNetwork.totalTokensThawing = graphNetwork.totalTokensThawing.plus(event.params.tokens)
+    graphNetwork.save()
+
+    provision.tokensThawing = provision.tokensThawing.plus(event.params.tokens)
+    provision.save()
+}
+
+export function handleTokensDeprovisioned(event: TokensDeprovisioned): void {
+    let graphNetwork = createOrLoadGraphNetwork(event.block.number, event.address)
+    let dataService = createOrLoadDataService(event.params.verifier)
+    let indexer = Indexer.load(event.params.serviceProvider.toHexString())!
+    let provision = createOrLoadProvision(event.params.serviceProvider, event.params.verifier, event.block.timestamp)
+
+    indexer.provisionedTokens = indexer.provisionedTokens.minus(event.params.tokens)
+    indexer.save()
+
+    dataService.totalTokensProvisioned = dataService.totalTokensProvisioned.minus(event.params.tokens)
+    dataService.save()
+
+    graphNetwork.totalTokensProvisioned = graphNetwork.totalTokensProvisioned.minus(event.params.tokens)
+    graphNetwork.save()
+
+    provision.tokensProvisioned = provision.tokensProvisioned.minus(event.params.tokens)
+    provision.save()
+}
+
+export function handleProvisionParametersSet(event: ProvisionParametersSet): void {
+    // To Do
+}
+
+export function handleProvisionParametersStaged(event: ProvisionParametersStaged): void {
+    // To Do
+}
+
+export function handleProvisionSlashed(event: ProvisionSlashed): void {
+    let graphNetwork = createOrLoadGraphNetwork(event.block.number, event.address)
+    let dataService = createOrLoadDataService(event.params.verifier)
+    let indexer = Indexer.load(event.params.serviceProvider.toHexString())!
+    let provision = createOrLoadProvision(event.params.serviceProvider, event.params.verifier, event.block.timestamp)
+
+    // Due to thawing tokens potentially getting cancelled, we will need to figure the thawing situation
+    indexer.provisionedTokens = indexer.provisionedTokens.minus(event.params.tokens)
+    indexer.stakedTokens = indexer.stakedTokens.minus(event.params.tokens)
+    indexer.save()
+
+    dataService.totalTokensProvisioned = dataService.totalTokensProvisioned.minus(event.params.tokens)
+    dataService.save()
+
+    graphNetwork.totalTokensProvisioned = graphNetwork.totalTokensProvisioned.minus(event.params.tokens)
+    graphNetwork.totalTokensStaked = graphNetwork.totalTokensStaked.minus(event.params.tokens)
+    graphNetwork.save()
+
+    provision.tokensProvisioned = provision.tokensProvisioned.minus(event.params.tokens)
```

Comment:
we also need to update `provision.tokensThawing`

Reply:
This one is actually a little bit weird, will discuss it on monday if possible